### ISSUE TITLE
Compiler reports `field is never read` even though it is used in an `impl`ementation

### ISSUE THREAD

**Issue by HenrikBach1** [rust-lang/rust#95105]
Hi

In my opinion, the below code forces the compiler to emit a false error on the field `seasonal_fruit` of `struct Breakfast`, which is apparently wrong, due to it is used in `impl Breakfast`:

```rust
mod back_of_house {
    pub struct Breakfast {
        pub toast: String,
        seasonal_fruit: String,
    }

    impl Breakfast {
        pub fn summer(toast: &str) -> Breakfast {
            Breakfast {
                toast: String::from(toast),
                seasonal_fruit: String::from("peaches"),
            }
        }
    }
}

pub fn eat_at_restaurant() {
    // Order a breakfast in the summer with Rye toast
    let mut meal = back_of_house::Breakfast::summer("Rye");
    // Change our mind about what bread we'd like
    meal.toast = String::from("Wheat");
    println!("I'd like {} toast please", meal.toast);

    // The next line won't compile if we uncomment it; we're not allowed
    // to see or modify the seasonal fruit that comes with the meal
    // meal.seasonal_fruit = String::from("blueberries");
}

fn main ()
{
    eat_at_restaurant()
}
```

The compiler emits the error:
```
[{
	"resource": "/projects/rust/paths-in-modules-7_3/src/main.rs",
	"owner": "rust-file:///projects/rust/paths-in-modules-7_3",
	"code": "dead_code",
	"severity": 4,
	"message": "field is never read: `seasonal_fruit`\n\nnote: `#[warn(dead_code)]` on by default",
	"source": "rustc",
	"startLineNumber": 4,
	"startColumn": 9,
	"endLineNumber": 4,
	"endColumn": 31
}]
```
The compiler that was used:
```
vscode ➜ /projects/rust/paths-in-modules-7_3 (master ✗) $ rustc --version
rustc 1.59.0 (9d1b2106e 2022-02-23)
```

Thank you
Henrik

**Comment by ThatsNoMoon** [comment]
The compiler is correct that it is never *read.* Using a field when constructing the struct is not reading it. If you, for example, printed its value, you would not get this error. As it is, `seasonal_fruit` is only ever written to.

**Comment by HenrikBach1** [comment]
Yes, that is a reasonable argument. I didn't think about this in that way; It is only used when the field is accessed in a read command. But for a newcomer like me, it is/was a little bit ackward in this thinking...

Thank you for your response.
